keps/0020-controller-overhaul.md

```diff
@@ -0,0 +1,160 @@
+
+This is by far the simplest solution but probably not the one with cleanest design. We just need an API call that will change the value of instance.Status.PlanStatus.Status to Pending for the status where instance.Status.PlanStatus.Name is equal to the plan you want to run (this is what current controller does when executing one of the built-in plans). Although it’s very easy to do this, it’s not very kubernetes idiomatic way of doing things especially because Status should never be updated from a client and it should just capture the state of the object.
+
+For some background, including definition of status subresource by [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status): `The status summarizes the current state of the object in the system, and is usually persisted with the object by an automated processes but may be generated on the fly. At some cost and perhaps some temporary degradation in behavior, the status could be reconstructed by observation if it were lost.`  
+
+#### Pros
+
+Very easy to implement considering the current state of the code
+
+#### Cons
+
+No way to start a plan with just pure kubectl - this is a reply from sig-cli channel where I posed that question (`It doesn't make sense to update the status manually, it should reflect the current state in the cluster. So no, you cannot set it though kubectl`)
+
+### Expose custom HTTP API from controller-manager to allow receiving commands
+
+Manager binary with controllers (that is run within the pod in the cluster) would in this case have to also expose HTTP API and we’ll have to expose a kubernetes service with this custom API. All clients (CLI/user scripts) will talk to this API every time they want to start a new plan. The API could be as simple as 
+
+HTTP POST `/start?planName=deploy`
+
+The validation whether a plan can be started or not would have to be part of the HTTP endpoint implementation.
+
+#### Pros
+
+The complexity of the implementation is still pretty managable and lower than having a full-featured sub-resource
+
+#### Cons
+
+Have to maintain a separate API (with API documentation)
+It is a non-kubernetes API, so no clients for various languages and no kubectl support (although if we keep the API that simple, almost a CURL can do the trick)
+More complex manager pod deployment with additional service
+How do we manage RBAC?
+
+### Have a separate CRD just for maintaining commands to execute a plan
+
+This would basically mean going half-way back to state pre KEP-18 where we had a separate CRD that was created for every plan that was executed. In this implementation we could have a much leaner CRD capturing only the requests to execute a plan with very lean status that does not duplicate the instance's status. Could be as simple as:
+
+```
+apiVersion: kudo.dev/v1beta1
+kind: PlanExecutionRequest
+Spec:
+  planName: deploy
+Status:
+  Started: true
```

Comment:
Probably more like:
```suggestion
  status: Started
```
to allow for more statuses like `Rejected`, `Aborted`, `Running` etc.

Reply:
yeah, good point. If we decide to go this way I definitely want to analyze this solution a bit further, so that we don't end up in a place again where status of a plan is in two separate CRDs and that's hard to update - which is not saying anything against your comment :) I think that is totally valid

Reply:
I guess we should separate here between the status of the plan execution **Request** (1) and status of the **plan** itself (2). 

**Plan**:
```
IN_PROGRESS
PENDING
COMPLETE
ERROR
...
```
 
**Request**:
```
Accepted
Rejected
Canceled// we'll inevitably come to forced plan executions at some point
???
```

P.S. we'll also need a reference to the instance in there and probably a timestamp.